Added ability to change resample_dt, path_tol etc from parameter server.

drewhamiltonasdf opened the pull request:
### Description

Hey all. I'm a bit of a noob. This is maybe second ever pull request so I'm not even sure if I'm doing it correctly. I just wanted to add the ability to change the Time Optimal Parameterization variables 'path_tolerance', 'resample_dt', and 'min_angle_change'. My approach may be naive here but it worked for my purposes. The only change is to add_time_optimal_parameterization.cpp'

This is a very simple change. Of course, I have no idea how to set up tests etc.

Best,
Drew Hamilton

### Checklist
- [ x] **Required by CI**: Code is auto formatted using [clang-format](http://moveit.ros.org/documentation/contributing/code)
- [ ] Extend the tutorials / documentation [reference](http://moveit.ros.org/documentation/contributing/)
- [ ] Document API changes relevant to the user in the [MIGRATION.md](https://github.com/ros-planning/moveit/blob/master/MIGRATION.md) notes
- [ ] Create tests, which fail without this PR [reference](https://ros-planning.github.io/moveit_tutorials/doc/tests/tests_tutorial.html)
- [ ] Include a screenshot if changing a GUI
- [ ] While waiting for someone to review your request, please help review [another open pull request](https://github.com/ros-planning/moveit/pulls) to support the maintainers

[//]: # "You can expect a response from a maintainer within 7 days. If you haven't heard anything by then, feel free to ping the thread. Thank you!"

simonschmeisser commented on moveit_ros/planning/planning_request_adapter_plugins/src/add_time_optimal_parameterization.cpp:
```diff
@@ -68,7 +107,8 @@ class AddTimeOptimalParameterization : public planning_request_adapter::Planning
       ROS_DEBUG("Running '%s'", getDescription().c_str());
-      TimeOptimalTrajectoryGeneration totg;
+      TimeOptimalTrajectoryGeneration totg(path_tolerance, resample_dt, min_angle_change);
+      //TimeOptimalTrajectoryGeneration totg;
```
please remove the "dead code"

drewhamiltonasdf replied:
got it.

simonschmeisser commented on moveit_ros/planning/planning_request_adapter_plugins/src/add_time_optimal_parameterization.cpp:
```diff
@@ -47,17 +47,56 @@ using namespace trajectory_processing;
   std::string getDescription() const override
   {
-    return "Add Time Optimal Parameterization";
+    return "Add Time Optimal Parameterization2";
```
I guess this was meant to be temporary

drewhamiltonasdf replied:
urf. yeah. I noticed that the second I submitted.

simonschmeisser commented on moveit_ros/planning/planning_request_adapter_plugins/src/add_time_optimal_parameterization.cpp:
```diff
@@ -47,17 +47,56 @@ using namespace trajectory_processing;
+
+    
+    double path_tolerance_override;
+    if (!nh.getParam("/path_tolerance", path_tolerance_override))
```
please move the param names to string constants (if you need to reuse them it's easier to avoid typos) and also note that other plugins omit the "/" in front. This reads the parameters relative to some position internal to the `NodeHandle` and allows to have multiple planning pipelines with different parameters

see here for an example:
 https://github.com/ros-planning/moveit/blob/253a2177f23c2d5e2e3fa415c75f720c9d6a98a5/moveit_ros/planning/planning_request_adapter_plugins/src/fix_start_state_collision.cpp#L61

drewhamiltonasdf replied:
Great. I'll give that a shot.

simonschmeisser reviewed (changes requested):
Thanks for your contribution! I found some stuff but it's all rather minor. Nevertheless people tend to be picky especially on first contributions as it's good to avoid any bad habits from the beginning :)

simonschmeisser commented:
Oh and don't use the word "hack" in commit messages ... it might stay there forever ;)

simonschmeisser commented on moveit_ros/planning/planning_request_adapter_plugins/src/add_time_optimal_parameterization.cpp:
```diff
@@ -47,12 +47,48 @@ using namespace trajectory_processing;
+  void initialize(const ros::NodeHandle& nh) override
   {
+
+    if (!nh.getParam(PATH_TOL_PARAM_NAME, path_tolerance))
```
Also this is minor as well, but if you switch the `if(! )` and the `else` it will be a tiny by simpler to read

felixvd commented:
The politically correct term is always „workaround“ :)

On Mon, Dec 14, 2020 at 17:30 Simon Schmeisser <notifications@github.com>
wrote:

> Oh and don't use the word "hack" in commit messages ... it might stay
> there forever ;)
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ros-planning/moveit/pull/2457#issuecomment-744271540>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABCTLOLRE4QT4IL5IRZJYBLSUXEJBANCNFSM4UZAZLGA>
> .
>

drewhamiltonasdf commented:
All style changes addressed. Modified fix_start_state_collision.cpp to match the more intuitive if-else, changed else statement to also use ROS_INFO_STREAM. I should mention that I'm neither a student nor an ambitious young programmer. Just a weekend warrior looking for a little more flexibility. I do appreciate all the comments though. Google has been my mentor. ;)

simonschmeisser reviewed (approved):
Thanks a lot for all the follow ups as well! Depending on what gets merged first, @felixvd or you will have to rebase a bit I fear when #2305 get's merged

drewhamiltonasdf commented:
I see a lot of red in the merge progress?

simonschmeisser commented:
There are unfortunately problems with travis (they removed their free offering and the paid version is still less powerful than what we used to get for free)

drewhamiltonasdf commented:
> There are unfortunately problems with travis (they removed their free offering and the paid version is still less powerful than what we used to get for free)

Ok. Just making sure there wasn't something I needed to fix on my end.

henningkayser commented on moveit_ros/planning/planning_request_adapter_plugins/src/add_time_optimal_parameterization.cpp:
```diff
@@ -47,12 +47,48 @@ using namespace trajectory_processing;
 {
 public:
+  
+  static const std::string RESAMPLE_PARAM_NAME;
```
I'd change these to `constexpr char PARAM[]` and move it out of this class and to the top of the surrounding namespace.

davetcoleman replied:
ping @drewhamiltonasdf

henningkayser reviewed (commented):
Very useful improvement, I'll approve this PR after the minor comment is addressed

davetcoleman commented on moveit_ros/planning/planning_request_adapter_plugins/src/add_time_optimal_parameterization.cpp:
```diff
@@ -47,12 +47,48 @@ using namespace trajectory_processing;
+
+    if (nh.getParam(PATH_TOL_PARAM_NAME, path_tolerance))
+    {
+      ROS_INFO_STREAM("Overriding path_tolerance parameters: %f", path_tolerance);
```
```suggestion
      ROS_INFO_STREAM_NAMED("add_time_optimal_parametrization","Overriding path_tolerance parameters: %f", path_tolerance);
```
By adding a namespace, users can turn on and off certain debug messages. Please apply this change to all the ROS_INFO_STREAMED below

davetcoleman commented on moveit_ros/planning/planning_request_adapter_plugins/src/add_time_optimal_parameterization.cpp:
```diff
@@ -79,8 +117,16 @@ class AddTimeOptimalParameterization : public planning_request_adapter::Planning
+  private:
+    double path_tolerance;
+    double resample_dt;
+    double min_angle_change;
```
@henningkayser shouldn't these member variables have "_" at end of name?
```suggestion
    double min_angle_change_;
```

henningkayser replied:
Yes, they should. It's even enforced by clang-tidy

davetcoleman reviewed (commented):
Great work!
